cartoframes/context.py

```diff
@@ -234,16 +233,89 @@ def read(self, table_name, limit=None, index='cartodb_id',
         # choose schema (default user - org or standalone - or shared)
         schema = 'public' if not self.is_org else (
             shared_user or self.creds.username())
-        query = 'SELECT * FROM "{schema}"."{table_name}"'.format(
+
+        table_columns = self._get_table_columns(table_name, schema)
+
+        query = self._get_read_query(table_name, schema, table_columns, limit)
+        result = self.copy_client.copyto_stream(query)
+        df = pd.read_csv(result)
+
+        return self._clean_dataframe_from_carto(df, table_columns, decode_geom)
+
+    def _get_read_query(self, table_name, schema, table_columns, limit=None):
+        """Create the read (COPY TO) query
+
+        Args:
+            table_name (str): Name of table in user's CARTO account.
+            schema (str): table schema name in user's CARTO account.
+            table_columns (dict): column names and types from a table.
+            limit (int, optional): Read only `limit` lines from
+                `table_name`. Defaults to ``None``, which reads the full table.
+
+        Returns:
+            str: COPY TO query
+        """
+        query_columns = list(table_columns.keys())
+        query_columns.remove('the_geom_webmercator')
+
+        query = 'SELECT {columns} FROM "{schema}"."{table_name}"'.format(
             table_name=table_name,
-            schema=schema)
+            schema=schema,
+            columns=', '.join(query_columns))
+
         if limit is not None:
             if isinstance(limit, int) and (limit >= 0):
                 query += ' LIMIT {limit}'.format(limit=limit)
             else:
                 raise ValueError("`limit` parameter must an integer >= 0")
 
-        return self.query(query, decode_geom=decode_geom)
+        return 'COPY ({query}) TO stdout WITH (FORMAT csv, HEADER true)'.format(query=query)
+
+    def _clean_dataframe_from_carto(self, df, table_columns, decode_geom=False):
+        """Clean a DataFrame with a dataset from CARTO:
+            - use cartodb_id as DataFrame index
+            - process date columns
+            - decode geom
+
+        Args:
+            df (pandas.DataFrame): DataFrame with a dataset from CARTO.
+            table_columns (dict): column names and types from a table.
+            decode_geom (bool, optional): Decodes CARTO's geometries into a
+              `Shapely <https://github.com/Toblerity/Shapely>`__
+              object that can be used, for example, in `GeoPandas
+              <http://geopandas.org/>`__.
+
+        Returns:
+            pandas.DataFrame
+        """
+        if 'cartodb_id' in df.columns:
+            df.set_index('cartodb_id', inplace=True)
+
+        for column_name in table_columns:
```

Comment:
Besides `date`s, are you sure that the schema will be faithfully represented from PG -> pandas.DataFrame? I know sometimes with postcodes/zipcodes leading zeros are removed when the column is erroneously converted to a numeric column instead of a string column.

Reply:
Since now we are using COPY and not the Import API, we are creating the table with the same schema as the Dataframe has (this is being done in the `write` method, still in CR). So when reading the table, the dataframe should have the same structure as it had before writing.

In summary, no more type/content guessing.

Reply:
- I have tested dates and works fine (we are doing the same in both cases)
- About the case of postcodes/zipcodes, I want to check if I understand it well: in db we have a string field with something like `01234` and in the DataFrame we should have the same string. Right? This case works fine too